tls/s2n_server_extensions.c

```diff
@@ -13,176 +13,56 @@
-        GUARD(s2n_extensions_server_key_share_send(conn, out));
-        /* Write cookie extension */
-        GUARD(s2n_extensions_cookie_send(conn, out));
-
-        return 0;
+    /* The ServerHello extension list size (uint16_t) is NOT written if the list is empty.
+     * This is to support older clients written before extensions existed that might fail
+     * on any unexpected bytes at the end of the ServerHello.
+     *
+     * This behavior is outlined in the TLS1.2 RFC: https://tools.ietf.org/html/rfc5246#appendix-A.4.1
+     *
+     * We COULD write the size if using TLS1.3, but let's prefer consistency and avoid
+     * altering existing behavior.
+     */
+    if(s2n_stuffer_data_available(out) - data_available_before_extensions == S2N_EMPTY_EXTENSION_LIST_SIZE) {
+        GUARD(s2n_stuffer_wipe_n(out, sizeof(uint16_t)));
     }
 
-    /* TLS 1.2 Extensions */
-
-    /* Write server name extension */
-    GUARD(s2n_server_extensions_server_name_send(conn, out));
-
-    /* Write kex extension */
-    GUARD(s2n_extension_send(&s2n_server_ec_point_format_extension, conn, out));
-    
-    /* Write the renegotiation_info extension */
-    GUARD(s2n_send_server_renegotiation_info_ext(conn, out));
-
-    /* Write ALPN extension */
-    GUARD(s2n_server_extensions_alpn_send(conn, out));
-
-    /* Write OCSP extension */
-    GUARD(s2n_server_extensions_status_request_send(conn, out));
-
-    /* Write Signed Certificate Timestamp extension */
-    GUARD(s2n_server_extensions_sct_list_send(conn, out));
-
-    /* Write max fragment length extension */
-    GUARD(s2n_server_extensions_max_fragment_length_send(conn, out));
-
-    /* Write session ticket extension */
-    GUARD(s2n_send_server_session_ticket_ext(conn, out));
-
-    return 0;
+    return S2N_SUCCESS;
 }
 
-int s2n_server_extensions_recv(struct s2n_connection *conn, struct s2n_blob *extensions)
+int s2n_server_extensions_recv(struct s2n_connection *conn, struct s2n_stuffer *in)
 {
-    struct s2n_stuffer in = {0};
-
-    GUARD(s2n_stuffer_init(&in, extensions));
-    GUARD(s2n_stuffer_write(&in, extensions));
-
-    while (s2n_stuffer_data_available(&in)) {
-        struct s2n_blob ext = {0};
-        uint16_t extension_type, extension_size;
-        struct s2n_stuffer extension = {0};
-
-        GUARD(s2n_stuffer_read_uint16(&in, &extension_type));
-        GUARD(s2n_stuffer_read_uint16(&in, &extension_size));
-
-        ext.size = extension_size;
-        ext.data = s2n_stuffer_raw_read(&in, ext.size);
-        notnull_check(ext.data);
+    s2n_parsed_extensions_list parsed_extension_list = { 0 };
+    GUARD(s2n_extension_list_parse(in, &parsed_extension_list));
 
-        GUARD(s2n_stuffer_init(&extension, &ext));
-        GUARD(s2n_stuffer_write(&extension, &ext));
+    /* Process supported_versions first so that we know which extensions list to use */
+    GUARD(s2n_extension_process(&s2n_server_supported_versions_extension, conn, &parsed_extension_list));
```

Comment:
nitpick: This comment is a little confusing to me because that extension isn't a given for 1.2 IIRC. I think for 1.2, we just use the default conn->server_highest_protocol_version (or some similarly named variable)